### PR TITLE
Feature/recurrent and multiple trainer MAPPO

### PR DESCRIPTION
**Pull request by DriesSmit**
## What?
Add recurrent training capabilities to MAPPO. Migrate MAPPO to now use the centralised variable source. This allows for training using multiple trainers.
## Why?
This update allows for the use of MAPPO in the recurrent training setting and with multiple trainers.
## How?
Adapted the MAPPO trainer to work in the recurrent setting. Added a recurrent executor to MAPPO. Migrated the old MAPPO code to use the new scaling version of Mava.
## Extra
.

### CODE REVIEW

File: mava/systems/tf/mappo/system.py [hunk 1]
```diff
@@ -50,28 +53,33 @@ def __init__(
...
         max_gradient_norm: Optional[float] = None,
-        max_queue_size: int = 100000,
-        batch_size: int = 256,
-        sequence_length: int = 10,
-        sequence_period: int = 5,
-        max_executor_steps: int = None,
+        max_queue_size: int = 1000,
+        batch_size: int = 32,
+        sequence_length: int = 20,
+        sequence_period: int = 10,
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/system.py:82]: Maybe we can set this to 19 / sequence_length-1 - https://github.com/deepmind/acme/blob/93a6a15b8c7b1d6a9c8099b9f2e9c693ab11a41e/acme/agents/jax/ppo/builder.py#L110 .

File: mava/systems/tf/mappo/training.py [hunk 2]
```diff
@@ -48,32 +49,29 @@ class MAPPOTrainer(mava.Trainer):
...
+        variable_client: VariableClient,
+        counts: Dict[str, Any],
         agent_net_keys: Dict[str, str],
-        checkpoint_minute_interval: int,
-        discount: float = 0.99,
-        lambda_gae: float = 1.0,
-        entropy_cost: float = 0.0,
-        baseline_cost: float = 1.0,
+        discount: float = 0.999,
+        lambda_gae: float = 1.0,  # Question (dries): What is this used for?
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/training.py:62]: Can we set the default to 0.95. This should be used in the advantage func calculation.

File: mava/systems/tf/mappo/training.py [hunk 3]
```diff
@@ -516,39 +540,37 @@ class CentralisedMAPPOTrainer(MAPPOTrainer):
...
         dataset: tf.data.Dataset,
         policy_optimizer: Union[snt.Optimizer, Dict[str, snt.Optimizer]],
         critic_optimizer: Union[snt.Optimizer, Dict[str, snt.Optimizer]],
+        variable_client: VariableClient,
+        counts: Dict[str, Any],
         agent_net_keys: Dict[str, str],
-        checkpoint_minute_interval: int,
-        discount: float = 0.99,
+        discount: float = 0.999,
         lambda_gae: float = 1.0,
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/training.py:553]: Please change the default to 0.95.

File: mava/systems/tf/mappo/training.py [hunk 4]
```diff
@@ -371,37 +383,41 @@ def _forward(self, inputs: Any) -> None:
...
-                policy_gradient_loss = tf.reduce_mean(
-                    -tf.minimum(
-                        tf.multiply(importance_ratio, normalized_gae),
-                        tf.multiply(clipped_importance_ratio, normalized_gae),
-                    ),
+
+                # Note (dries): Maybe add this in again? But this might be breaking
+                # training. mean, variance = tf.nn.moments(gae, axes=[0, 1],
+                # keepdims=True)
+                # normalized_gae = (gae - mean) / tf.sqrt(variance)
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/training.py:390]: Was there evidence this was breaking training / hurting performance? Maybe we just need to make sure the denominator is not 0 , by doing this ` normalized_gae = (gae - mean) / (tf.sqrt(variance)+ 1e-8)` .

File: mava/systems/tf/mappo/training.py [hunk 5]
```diff
@@ -574,3 +595,64 @@ def _get_critic_feed(
...
+        policy_networks: Dict[str, snt.Module],
+        critic_networks: Dict[str, snt.Module],
+        dataset: tf.data.Dataset,
+        policy_optimizer: Union[snt.Optimizer, Dict[str, snt.Optimizer]],
+        critic_optimizer: Union[snt.Optimizer, Dict[str, snt.Optimizer]],
+        counts: Dict[str, Any],
+        variable_client: VariableClient,
+        agent_net_keys: Dict[str, str],
+        discount: float = 0.999,
+        lambda_gae: float = 1.0,
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/training.py:616]: Please set this 0.95 as a default.

**Review by KaleabTessera — COMMENTED** [review]
Thanks so much @DriesSmit :open_hands: This is really great! Just see my minor comments.

File: mava/systems/tf/mappo/builder.py [hunk 6]
```diff
@@ -177,10 +245,19 @@ def make_dataset_iterator(
         """
 
         # Create tensorflow dataset to interface with reverb
-        dataset = datasets.make_reverb_dataset(
+        # dataset = datasets.make_reverb_dataset(
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/builder.py:248]: Could we please clean up the comments.

File: mava/systems/tf/mappo/builder.py [hunk 7]
```diff
@@ -177,10 +245,19 @@ def make_dataset_iterator(
...
+        #     server_address=replay_client.server_address,
+        #     batch_size=self._config.batch_size,
+        # )
+
+        dataset = reverb.TrajectoryDataset.from_table_signature(
             server_address=replay_client.server_address,
-            batch_size=self._config.batch_size,
-        )
+            table=table_name,
+            max_in_flight_samples_per_worker=1,
```
Comment (KaleabTessera) [inline mava/systems/tf/mappo/builder.py:256]: Might be a good idea to set this to 2 * batch size following acme - https://github.com/deepmind/acme/blob/93a6a15b8c7b1d6a9c8099b9f2e9c693ab11a41e/acme/agents/jax/ppo/builder.py#L97 .

File: mava/systems/tf/maddpg/execution.py [hunk 8]
```diff
@@ -402,12 +402,11 @@ def observe_first(
             self._net_keys_to_ids,
         )
 
-        if self._store_recurrent_state:
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/maddpg/execution.py:405]: This option is now always true. To me, it did not make sense to not always store the recurrent states. Or do you disagree?

File: mava/systems/tf/maddpg/execution.py [hunk 9]
```diff
@@ -430,12 +429,11 @@ def observe(
             return
         _, policy = actions
 
-        if self._store_recurrent_state:
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/maddpg/execution.py:433]: See above comment.

File: mava/systems/tf/mappo/builder.py [hunk 10]
```diff
@@ -75,13 +76,19 @@ class MAPPOConfig:
...
+    network_sampling_setup: List
+    net_keys_to_ids: Dict[str, int]
+    unique_net_keys: List[str]
     checkpoint_minute_interval: int
     sequence_length: int = 10
     sequence_period: int = 5
     discount: float = 0.99
     lambda_gae: float = 0.95
-    max_queue_size: int = 100_000
+    max_queue_size: int = 1000
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/mappo/builder.py:91]: Because of the exact reason you pointed out in your link :smile: `This (large queue size) will not work for distributed setup
    # if learner is slower than actors (the queue will fill up and the learner
    # will be offpolicy).`

File: mava/systems/tf/mappo/execution.py [hunk 11]
```diff
@@ -83,96 +98,104 @@ def _policy(
...
 
         # Compute the policy, conditioned on the observation.
-        policy = self._policy_networks[network_key](observation)
+        logits = self._policy_networks[agent_key](batched_observation)
 
         # Sample from the policy and compute the log likelihood.
-        action = policy.sample()
-        log_prob = policy.log_prob(action)
+        action = tfd.Categorical(logits).sample()
+        # action = tf2_utils.to_numpy_squeeze(action)
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/mappo/execution.py:120]: Cleaned.

File: mava/systems/tf/mappo/execution.py [hunk 12]
```diff
@@ -83,96 +98,104 @@ def _policy(
...
 
         # Cast for compatibility with reverb.
         # sample() returns a 'int32', which is a problem.
-        if isinstance(policy, tfp.distributions.Categorical):
-            action = tf.cast(action, "int64")
+        # if isinstance(policy, tfp.distributions.Categorical):
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/mappo/execution.py:124]: Cleaned.

File: mava/systems/tf/mappo/execution.py [hunk 13]
```diff
@@ -192,20 +215,242 @@ def observe(
...
+        # Compute the policy, conditioned on the observation.
+        logits, new_state = self._policy_networks[agent_key](batched_observation, state)
+
+        # Sample from the policy and compute the log likelihood.
+        action = tfd.Categorical(logits).sample()
+        # action = tf2_utils.to_numpy_squeeze(action)
+
+        # Cast for compatibility with reverb.
+        # sample() returns a 'int32', which is a problem.
+        # if isinstance(policy, tfp.distributions.Categorical):
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/mappo/execution.py:320]: Cleaned.

File: mava/systems/tf/mappo/networks.py [hunk 14]
```diff
@@ -109,7 +146,7 @@ def make_default_networks(
                     networks.MultivariateNormalDiagHead(
                         num_dimensions=num_actions,
                         w_init=tf.initializers.VarianceScaling(1e-4, seed=seed),
-                        b_init=tf.initializers.Zeros(seed=seed),
+                        b_init=tf.initializers.Zeros(),
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/mappo/networks.py]: If I remember correctly it is because it does not require a seed value. Zeros remain zeros. The variance one does, but not a zero bias term.

File: mava/systems/tf/mappo/training.py (hunk 2 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/training.py:62]: Yes, this is used in the Generalized Return Estimation. But I think we are doing a double discount in the `trfl.td_lambda` function if we set `lambda_gae` not equal to 1.0. We should probably investigate this again.

File: mava/systems/tf/mappo/training.py (hunk 3 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/training.py:553]: See above comment.

File: mava/systems/tf/mappo/training.py (hunk 5 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/training.py:616]: Please see above comment.

File: mava/systems/tf/mappo/builder.py (hunk 6 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/builder.py:248]: Fixed.

File: mava/systems/tf/mappo/training.py (hunk 4 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/training.py:390]: I would have to investigate this again. I just remembered that this caused NaNs. I cannot remember if I tried adding a small constant below the division. It's definitely a good idea yes:+1:

File: mava/systems/tf/mappo/system.py (hunk 1 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/system.py:82]: Interesting :slightly_smiling_face: I will have to look into this. Thanks for pointing this out.

File: mava/systems/tf/mappo/builder.py (hunk 7 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/builder.py:257]: Good point. Do you know what `max_in_flight_samples_per_worker` represents?

File: mava/systems/tf/maddpg/execution.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (KaleabTessera) [reply mava/systems/tf/maddpg/execution.py:405]: I guess this might not be necessary. Is there a situation we would use the rec executor, without storing the recurrent state ? :thinking:

File: mava/systems/tf/mappo/builder.py (hunk 7 above)
Comment (KaleabTessera) [reply mava/systems/tf/mappo/builder.py:257]: Found the readme here - https://github.com/deepmind/reverb/blob/b57722c77cb5f36f82688aae8502c23a7a8de6bd/reverb/dataset.py#L76 .

File: mava/systems/tf/maddpg/execution.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/systems/tf/maddpg/execution.py:405]: Not that I know of.

File: mava/systems/tf/mappo/system.py (hunk 1 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/system.py:82]: Done.

File: mava/systems/tf/mappo/training.py (hunk 2 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/training.py:62]: I changed it to 0.95. Although I am still not sure why it should just not always equal 1.0.

File: mava/systems/tf/mappo/builder.py (hunk 7 above)
Comment (DriesSmit) [reply mava/systems/tf/mappo/builder.py:257]: Fixed.

File: examples/tf/debugging/simple_spread/recurrent/state_based/run_mappo.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    "mava_id",
+    str(datetime.now()),
+    "Experiment identifier that can be used to continue experiments.",
+)
+flags.DEFINE_string("base_dir", "~/mava", "Base dir to store experiments.")
+
+
+def main(_: Any) -> None:
+
+    recurrent_test = False
```
Comment (arnupretorius) [inline examples/tf/debugging/simple_spread/recurrent/state_based/run_mappo.py:55]: Should we remove these booleans and fix it to be recurrent?

File: mava/components/tf/architectures/centralised.py [hunk 16]
```diff
@@ -143,11 +144,12 @@ def _get_critic_specs(
 
         for agent_type, agents in agents_by_type.items():
             agent_key = agents[0]
+            net_key = self._agent_net_keys[agent_key]
 
             # TODO (dries): Add a check to see if all
             #  self._embed_specs[agent_key].shape are of the same shape
```
Comment (arnupretorius) [inline mava/components/tf/architectures/centralised.py:150]: Remove?

File: mava/components/tf/architectures/state_based.py [hunk 17]
```diff
@@ -98,11 +99,26 @@ def _get_critic_specs(
...
+        critic_obs_spec = []
+        for spec in critic_env_state_spec:
+            critic_obs_spec.append(
+                tf.TensorSpec(
+                    shape=spec.shape,
+                    dtype=tf.dtypes.float32,
+                )
+            )
+
+        # TODO (dries): Fix this
```
Comment (arnupretorius) [inline mava/components/tf/architectures/state_based.py:118]: Still relevant?

File: mava/components/tf/architectures/state_based.py [hunk 18]
```diff
@@ -148,6 +165,61 @@ def __init__(
...
```
Comment (arnupretorius) [inline mava/components/tf/architectures/state_based.py:211]: Remove?

**Review by arnupretorius — APPROVED** [review]
Thanks @DriesSmit! 👍 Did a quick smoke review. Left minor comments. Happy with the benchmarking on this and will be useful for comparing with Jax systems.

File: mava/utils/sort_utils.py [hunk 19]
```diff
@@ -27,6 +27,7 @@ def sample_new_agent_keys(
     agents: List,
     network_sampling_setup: List,
     net_keys_to_ids: Dict[str, int] = None,
+    fix_sampler: Optional[List] = None,
```
Comment (KaleabTessera) [inline mava/utils/sort_utils.py:30]: @DriesSmit Can we document what this does ?

**Review by KaleabTessera — DISMISSED** [review]
Thanks @DriesSmit , great work on getting this in :fire: :open_hands: :fire: 

Just a minor commont on adding doc strings for ` fix_sampler`.

File: examples/tf/debugging/simple_spread/recurrent/state_based/run_mappo.py (hunk 15 above)
Comment (DriesSmit) [reply examples/tf/debugging/simple_spread/recurrent/state_based/run_mappo.py:55]: Fixed.

File: mava/components/tf/architectures/centralised.py (hunk 16 above)
Comment (DriesSmit) [reply mava/components/tf/architectures/centralised.py:150]: Fixed.

File: mava/components/tf/architectures/state_based.py (hunk 17 above)
Comment (DriesSmit) [reply mava/components/tf/architectures/state_based.py:118]: I updated the message.

File: mava/components/tf/architectures/state_based.py (hunk 18 above)
Comment (DriesSmit) [reply mava/components/tf/architectures/state_based.py:211]: I updated the message.

File: mava/utils/sort_utils.py (hunk 19 above)
Comment (DriesSmit) [reply mava/utils/sort_utils.py:30]: Added it.

**Review by arnupretorius — APPROVED** [review]
Thanks Dries! 🔥

**Review by KaleabTessera — APPROVED** [review]
